File: backend/apps/academics/api/serializers/promote.py
from rest_framework import serializers


class PromoteExecuteSerializer(serializers.Serializer):
    from_session_id = serializers.IntegerField()
    from_class_id = serializers.IntegerField()
    from_section_id = serializers.IntegerField()
    to_session_id = serializers.IntegerField()
    to_class_id = serializers.IntegerField()
    to_section_id = serializers.IntegerField()
    to_subject_group_id = serializers.IntegerField(required=False, allow_null=True)
    student_ids = serializers.ListField(
        child=serializers.IntegerField(), required=False, allow_empty=True
    )
    deactivate_source = serializers.BooleanField(required=False, default=True)
    mark_alumni = serializers.BooleanField(required=False, default=False)

File: frontend/src/types/academics/promote.ts
export interface PromotePreviewStudent {
  student_id: number;
  student_session_id: number;
  admission_no: string | null;
  name: string | null;
  current_class_name: string | null;
  current_section_name: string | null;
  eligible: boolean;
  blockers: string[];
  fee_warning: boolean;
}

export interface PromotePreview {
  eligible_count: number;
  already_in_target_count: number;
  inactive_skipped_count: number;
  students: PromotePreviewStudent[];
  warnings: string[];
}

export interface PromotePreviewParams {
  from_session_id: number;
  from_class_id: number;
  from_section_id: number;
  to_session_id: number;
  to_class_id: number;
  to_section_id: number;
}

export interface PromoteExecutePayload extends PromotePreviewParams {
  to_subject_group_id?: number | null;
  student_ids?: number[];
  deactivate_source?: boolean;
  mark_alumni?: boolean;
}

export interface PromoteExecuteResult {
  promoted_count: number;
  skipped_count: number;
  promoted: Array<{
    student_id: number;
    new_student_session_id: number;
  }>;
}

export interface PromoteWizardState {
  fromSessionId?: number;
  fromClassId?: number;
  fromSectionId?: number;
  toSessionId?: number;
  toClassId?: number;
  toSectionId?: number;
  toSubjectGroupId?: number | null;
  deactivateSource: boolean;
  markAlumni: boolean;
  selectedStudentIds: number[];
}
